File: sdks/kotlin/core/src/main/kotlin/io/getstream/visionagents/core/VisionAgents.kt
package io.getstream.visionagents.core

import okhttp3.OkHttpClient

/**
 * The router, as a phone sees it.
 *
 * Two lines get a conversation going:
 *
 *     val agents = VisionAgents(url = "https://your-router", customerId = "acme")
 *     val chat = agents.agent("docs").chat()
 *
 * Opening a conversation, reading back its turns, going back to one of them, branching off and
 * ending it is the whole of what is here, because it is the whole of what the router lets a
 * device do. Configuring an agent and minting a token to join a call are server-side only: they
 * belong to a backend, which has the Go, Python or Node SDK and hands down what the app needs.
 */
public class VisionAgents(public val backend: Backend) : AutoCloseable {
    public constructor(
        url: String,
        customerId: String = "",
        apiKey: String = "",
        okHttpClient: OkHttpClient? = null,
    ) : this(Backend(url, customerId, apiKey, okHttpClient))

    /** Every conversation this caller has, whichever agent holds it. */
    public val sessions: Sessions = Sessions(backend)

    /** Says who this device is acting for, and how to prove it. See [Backend.setUser]. */
    public fun setUser(user: User, token: TokenProvider): Unit = backend.setUser(user, token)

    public fun setUser(user: User, token: String): Unit = backend.setUser(user, token)

    /** Acts as this guest from here on. */
    public fun setUser(guest: Guest): Unit = backend.setUser(guest)

    /** One agent, by the name its config was synced under. */
    public fun agent(name: String): Agent = Agent(this, name)

    /** Looking something up, under a stored router config or none. */
    public fun router(config: String = "", tags: Map<String, String> = emptyMap()): Router =
        Router(backend, config, tags)

    /**
     * Holds a conversation in writing: no call is joined, nothing is transcribed or spoken.
     *
     * The replies still come through the model with the same instructions, skills and knowledge
     * a call would have had, and stream into [AgentSession.conversation]. It returns with the
     * socket open, so the tools are already being answered.
     */
    public suspend fun chat(options: SessionOptions = SessionOptions()): AgentSession =
        follow(sessions.create(checked(options)), options.tools)

    /**
     * Puts an agent on a call and follows it.
     *
     * The agent joins as soon as this returns. Joining the same call from this device is what
     * the rtc module is for; this only starts the agent and gives you the state.
     */
    public suspend fun voice(callId: String, options: SessionOptions = SessionOptions()): AgentSession =
        follow(sessions.create(checked(options), callId), options.tools)

    /**
     * Follows a session this caller already has open, without creating one.
     *
     * Use this after a relaunch, on another screen, or for a fork. A session opened by somebody
     * else is not found, because reading one is reading a conversation.
     */
    public suspend fun attach(sessionId: String, tools: List<AgentTool> = emptyList()): AgentSession =
        follow(sessions.get(sessionId), tools)

    /**
     * Gets or creates a guest, so somebody can talk to an agent before they sign up.
     *
     * Remembered in [store] where there is one, so the next launch is the same guest and not a
     * second one with an empty history. Pass it to [setUser] to act as them. Moving a guest's
     * conversations onto the account they turn out to be is server-side only: it is the backend
     * that just authenticated them that knows which guest they were.
     */
    public suspend fun guestUser(options: GuestOptions = GuestOptions(), store: GuestStore? = null): Guest =
        backend.guest(options, store)

    /** Forgets the remembered guest without minting another, which is what signing out is. */
    public fun forgetGuest(store: GuestStore) {
        store.clear()
        backend.clearUser()
    }

    override fun close() {
        backend.close()
    }

    private suspend fun follow(session: Session, tools: List<AgentTool>): AgentSession =
        AgentSession(backend, session, tools).also { it.start() }

    private fun checked(options: SessionOptions): SessionOptions {
        val duplicate = options.tools.groupBy { it.name }.entries.firstOrNull { it.value.size > 1 }
        if (duplicate != null) {
            throw AgentsException.Configuration("${duplicate.key} is declared twice")
        }
        return options
    }
}

/**
 * One agent, and the conversations held with it.
 *
 * The name is the one the agent's config was synced under, which the router resolves; a name
 * that matches nothing is refused rather than starting an agent with no config.
 */
public class Agent internal constructor(private val agents: VisionAgents, public val name: String) {
    /** This agent's conversations: opening one names it, and listing is narrowed to it. */
    public val sessions: Sessions = Sessions(agents.backend, name)

    /** Holds a conversation with this agent in writing. See [VisionAgents.chat]. */
    public suspend fun chat(options: SessionOptions = SessionOptions()): AgentSession =
        agents.chat(options.copy(agent = options.agent ?: name))

    /** Holds a conversation with this agent in writing, answering these tools here. */
    public suspend fun chat(vararg tools: AgentTool): AgentSession =
        chat(SessionOptions(tools = tools.toList()))

    /** Puts this agent on a call and follows it. See [VisionAgents.voice]. */
    public suspend fun voice(callId: String, options: SessionOptions = SessionOptions()): AgentSession =
        agents.voice(callId, options.copy(agent = options.agent ?: name))
}

File: sdks/dart/core/lib/src/vision_agents.dart
import 'package:http/http.dart' as http;

import 'agent_session.dart';
import 'backend.dart';
import 'convert.dart';
import 'errors.dart';
import 'generated/api.dart' as api;
import 'guests.dart';
import 'models.dart';
import 'router.dart';
import 'sessions.dart';
import 'tools.dart';
import 'wire.dart';

/// The router, as a device sees it.
///
/// Two lines get a conversation going:
///
///     final agents = VisionAgents(url: Uri.parse('https://your-router'), customerId: 'acme');
///     final chat = await agents.agent('docs').chat();
///
/// Opening a conversation, following it, reading back its turns, going back to one of them,
/// branching off and ending it, a guest to do it as, and looking something up are the whole
/// of what is here, because they are the whole of what the router lets a device do. What an
/// agent is configured as and a token to join a call with are server-side only: they belong
/// to the app's backend, which hands down what the app needs.
///
/// The constructor does no I/O.
final class VisionAgents {
  VisionAgents({
    required Uri url,
    String customerId = '',
    String apiKey = '',
    String userId = '',
    TokenProvider? token,
    http.Client? httpClient,
  }) : this.withBackend(
         Backend(url: url, customerId: customerId, apiKey: apiKey, userId: userId, token: token),
         httpClient: httpClient,
       );

  /// An injected [httpClient] is the caller's and is never closed here.
  VisionAgents.withBackend(this.backend, {http.Client? httpClient})
    : _client = httpClient ?? http.Client(),
      _ownsClient = httpClient == null {
    _operations = api.Operations(HttpWire(backend, _client));
    sessions = Sessions(_operations);
  }

  final Backend backend;
  final http.Client _client;
  final bool _ownsClient;
  late final api.Operations _operations;

  /// This caller's conversations, whichever agent held them.
  late final Sessions sessions;

  /// One agent, by the name its config was stored under.
  Agent agent(String name) => Agent._(this, name);

  /// The same router, asked by somebody else. Shares this one's HTTP client.
  VisionAgents withUser(String userId, {TokenProvider? token}) =>
      VisionAgents.withBackend(backend.withUser(userId, token: token), httpClient: _client);

  /// The same router, asked by a guest. On a router verifying tokens, the guest's own token is
  /// the credential.
  VisionAgents withGuest(GuestUser guest) =>
      withUser(guest.id, token: backend.apiKey.isEmpty ? null : () async => guest.token);

  /// Holds a conversation in writing: no call is joined, nothing is transcribed or spoken.
  ///
  /// The replies still come from the model with the same instructions, skills and knowledge a
  /// call would have, and arrive as deltas on the socket, which is already open when this
  /// returns.
  Future<AgentSession> chat([SessionOptions options = const SessionOptions()]) =>
      _follow(sessions, options, null);

  /// Puts an agent on a call and follows it.
  ///
  /// The agent joins as soon as this returns. Joining the same call from this device is what
  /// the rtc package is for; this only starts the agent and gives you the state.
  Future<AgentSession> voice(String callId, [SessionOptions options = const SessionOptions()]) =>
      _follow(sessions, options, callId);

  /// Follows a session this caller already has, without creating one: after a relaunch, on
  /// another screen, or a fork.
  ///
  /// A session opened by somebody else is not found, because reading one is reading a
  /// conversation.
  Future<AgentSession> attach(String sessionId, {List<AgentTool> tools = const []}) async {
    final session = AgentSession(
      await sessions.get(sessionId),
      backend: backend,
      sessions: sessions,
      tools: tools,
    );
    await session.start();
    return session;
  }

  /// Looking something up, under a stored router config or none.
  SearchRouter router({String config = '', Map<String, String> tags = const {}}) =>
      SearchRouter(_operations, config: config, tags: tags);

  /// Gets or creates a guest, so somebody can talk to an agent before they sign up.
  ///
  /// Remembered in [store] so coming back is the same guest, with the same conversations,
  /// rather than a second one with an empty history. A remembered guest whose token has
  /// expired is minted a fresh token under the same id. [fresh] mints a new guest whatever is
  /// remembered, for a "not me" button: reusing the remembered one would hand the person at
  /// the keyboard somebody else's conversations.
  ///
  /// Throws a [RouterException] with status 403 when the app does not admit guests.
  Future<GuestUser> guestUser({
    String? name,
    Map<String, Object?>? custom,
    GuestStore? store,
    bool fresh = false,
  }) async {
    final held = store == null || fresh ? null : await readGuest(store);
    final expires = held?.expiresAt;
    if (held != null && (expires == null || expires.isAfter(DateTime.now().add(_leeway)))) {
      return held;
    }
    final minted = guestOf(
      await _operations.createGuestUser(
        body: api.GuestUserRequest(id: held?.id, name: name, custom: custom),
      ),
    );
    if (store != null) {
      await writeGuest(store, minted);
    }
    return minted;
  }

  /// Forgets the remembered guest without minting another, which is what signing out does.
  Future<void> forgetGuest(GuestStore store) => store.clear();

  /// Closes the HTTP client, if this created it. Sessions close on their own.
  void close() {
    if (_ownsClient) {
      _client.close();
    }
  }

  Future<AgentSession> _follow(Sessions sessions, SessionOptions options, String? callId) async {
    final session = AgentSession(
      await sessions.create(options, callId),
      backend: backend,
      sessions: sessions,
      tools: options.tools,
    );
    try {
      await session.start();
    } on AgentsException {
      // The agent is in the conversation whether or not anybody here can follow it, so it is
      // ended rather than left holding a call nobody is listening to.
      try {
        await sessions.close(session.id);
      } on AgentsException {
        // Reporting why the socket did not open matters more than this.
      }
      rethrow;
    }
    return session;
  }
}

/// A token about to expire is as good as expired by the time it reaches the router.
const _leeway = Duration(minutes: 1);

/// One agent: its conversations, and holding a new one.
///
/// The JavaScript SDK's `client.agent("docs")`: every session opened or listed here is this
/// agent's, unless the options name another config.
final class Agent {
  Agent._(this._agents, this.name) : sessions = Sessions(_agents._operations, agent: name);

  final VisionAgents _agents;

  /// The name its config was stored under.
  final String name;

  /// This agent's conversations.
  final Sessions sessions;

  /// Holds a conversation with this agent in writing.
  Future<AgentSession> chat([SessionOptions options = const SessionOptions()]) =>
      _agents._follow(sessions, options, null);

  /// Puts this agent on a call and follows it.
  Future<AgentSession> voice(String callId, [SessionOptions options = const SessionOptions()]) =>
      _agents._follow(sessions, options, callId);
}
